### PR TITLE
Adding `preview` deployment feature

### PR DESCRIPTION
**Pull request by vhvb1989**
fix: https://github.com/Azure/azure-dev/issues/141

### with bicep, fresh env:

image

### running `azd provision` after `azd provision --preview` :

image

### running with `--preview` after provision:

image

### CODE REVIEW

**Review by savannahostrowski — CHANGES_REQUESTED** [review]
Overall, this looks pretty good to me! I'll defer to @Austinauth on styling, colour choices for states etc.

Just a couple of copy updates.

**Review by savannahostrowski — APPROVED** [review]

**Comment by Austinauth** [comment]
Hey Victor, this is looking good, appreciate the format change!

Not a stop ship, but my only minor nit is that for the normal progress list we have only have a single space between the `status:` `service:` and `identifier`. In your mockups you've spaced things out to display things in more of a table format. 
I actually really like your approach, but it's not consistent with the existing pattern. I'll make note and in the future we can revisit to see if we can arrive at a consistent pattern that works for both.

File: cli/azd/pkg/infra/provisioning/terraform/terraform_provider.go
Comment (ellismg) [inline cli/azd/pkg/infra/provisioning/terraform/terraform_provider.go:205]: I would have expected that if I ran `azd provision --what-if` when Terraform was my IaC provider this would just print the output from `tf plan`.

I don't think that's what is going to happen today with the changes as is.

I'm not 100% sure what exactly we want to do here - I think we want to just end up showing the tf plan output as is (that matches how we handle showing the raw output from terraform when we do a provision), but it's something we should think about.

I also wonder if on the provider interface we should merge `Plan` and `WhatIfDeploy` somehow.  I think the problem is that unlike terraform, where you always compute a plan before applying it, in ARM land you can do a deploy without a what-if operation, so we'd have to figure out a way to communicate into Plan for the ARM case if you actually wanted to do the what-if part or not.

**Comment by ellismg** [comment]
From a styling point of view, I sort of wondering if re-using green for "no change" is the right move.  I'm slightly nervous that it's the same color as "creating a resource" but these are like very different operations.  I wonder if there's some grey or something we could be using here instead.  Maybe something for @Austinauth to noodle on.

**Comment by ellismg** [comment]
Question for you, @vhvb1989, In the wonderful screenshots that you show, I'm guessing you ran `azd provision --what-if` with no changes after the initial provision and so in theory we would expect to see "NoChanges" for all the resources, right?

Do you know what about these resources the What If endpoint thinks is going to change?  I understand that we are sort of at the mercy of the control plane here telling us about the diffs, and so maybe we are always going to be in this state until we can work with the RP providers to improve the diffing logic, but we had talked about some sort of `--details` switch that might show for resources which are changing what exactly about them is changing.

Just wondering if you've looked at the raw results coming back from the API and have a better understanding of what is changing here.

File: cli/azd/pkg/infra/provisioning/terraform/terraform_provider.go
Comment (vhvb1989) [reply cli/azd/pkg/infra/provisioning/terraform/terraform_provider.go:205]: > I would have expected that if I ran azd provision --what-if when Terraform was my IaC provider this would just print the output from tf plan.

That's what it is currently happening due to the provision-manager always calling `Pan()` before `Deploy()` . This makes terraform to only run `Plan()` and show that output. (I ran a test).  Since it returns an empty list of changes, azd does't add more output after it is donde.

I don't think merging Plan() and what-if is a good idea.  I see the `bicep-planning` as the actual `bicep build` operation (transform to arm). 

How does it work in the case of Pulumi?

File: cli/azd/cmd/testdata/TestUsage-azd-provision.snap
Comment (rajeshkamal5050) [inline cli/azd/cmd/testdata/TestUsage-azd-provision.snap:16]: `--what-if` feels bicep centric given that we will also adopt TF's plan eventually. 

Any other options like `preview` `dry-run` etc?

@ellismg @savannahostrowski @vhvb1989

Comment (savannahostrowski) [reply cli/azd/cmd/testdata/TestUsage-azd-provision.snap:16]: I was also thinking about this. I'd probably lean toward `--preview` over `--dry-run`

File: cli/azd/pkg/infra/provisioning/terraform/terraform_provider.go
Comment (ellismg) [reply cli/azd/pkg/infra/provisioning/terraform/terraform_provider.go:205]: Thanks for the explantation.  I'm still a little nervous about the fact that we have a thing on our interface that is specific to one provider, but that's something we can noodle on in the background and think if there's some better design here.  I think what you have here is good for now, so this feedback is not blocking.

> How does it work in the case of Pulumi?

Pulumi works similar-ish to Terraform.  The typical workflow there is to do a `pulumi up` which will run a preview, ask you to accept it, and then apply it.  There's also a `--skip-preview` flag you can pass, and and an explicit `preview` command that just runs the planning step.

Unlike Terraform, the default workflow doesn't encourage you to produce a reified plan that you then apply later (although they now support something like this).

I suspect that if we added Pulumi support, `Plan()` would call `pulumi preview`, `Deploy()` would call `pulumi --skip-preview`, and then `WhatIfDeploy()` would be a no-op here.

File: cli/azd/cmd/testdata/TestUsage-azd-provision.snap
Comment (ellismg) [reply cli/azd/cmd/testdata/TestUsage-azd-provision.snap:16]: I like `--preview` as well and it matches the copy that @vhvb1989 used in the user facing messages (i.e. "Preview Azure resource changes").

File: cli/azd/pkg/infra/provisioning/terraform/terraform_provider.go
Comment (vhvb1989) [reply cli/azd/pkg/infra/provisioning/terraform/terraform_provider.go:205]: I see.
Now I agree with you.  There should be only `plan` and `deploy`.  
That requires updating the `provision` model to only call `plan()` when `--preview` flag.  Then moving the actual `plan()` implementations from the providers to be part of `Deploy()` .
Then making `plan()` to be the implementation I added in this PR.

I think it is worth it to have that in a new PR. 
I am happy to do it (I think it's straight forward).  I just think it will be easier to review and see what's changing in a new PR than making this PR to handle both changes

File: cli/azd/cmd/util.go
Comment (ellismg) [inline cli/azd/cmd/util.go:249]: I wonder if long term we would want to remove this message from `--preview`.  I'm not sure pointing folks at the RG in this case every time provides a lot of value.  Perhaps we have have an `azd show` targeted at humans and not computers (e.g. the `--output=json` mode that we support today for editors) we can remove this.

Just a note - non blocking feedback.

File: cli/azd/pkg/output/ux/preview_provision.go
Comment (ellismg) [inline cli/azd/pkg/output/ux/preview_provision.go:21]: Wondering if we should implement [`fmt.Stringer`](https://pkg.go.dev/fmt#Stringer) on this type, so we could display `ChangeTypeNoChange` as "Unchanged" instead of "NoChange".  This is the one of these that ends up not being an English word, and having the text smushed together like this makes the output feel a little more "computer'y" and less friendly.

**Review by ellismg — APPROVED** [review]
Great Job here, @vhvb1989.  Really happy with how this all turned out.

I'm fine if you want to land this and do the changes to the Provider interface later, whatever works better for you.

**Pull request merged**
